theano/tests/unittest_tools.py

```diff
@@ -311,8 +311,9 @@ def str_diagnostic(expected, value, rtol, atol):
         pass
     # Use the same formula as in _allclose to find the tolerance used
     narrow = 'float32', 'complex64'
-    if ((str(expected.dtype) in narrow) or
-        (str(value.dtype) in narrow)):
+    # If the user passed a list with tensor, dtype won't be defined
+    if ((str(getattr(expected, 'dtype', '')) in narrow) or
+        (str(getattr(value, 'dtype', '')) in narrow)):
```

Comment:
This will conflict with my changes for float16 since I remove this part to instead call the same code as in _allclose.